File: Python/textbook_track/chapter06/ba6f.py
# Implement ChromosomeToCycle
import sys

from enum import Enum


class Sign(Enum):
    PLUS = '+'
    MINUS = '-'


class SignedNumber:
    def __init__(self, sign, number):
        self._sign = sign
        self._number = number

    def __repr__(self):
        return f'{self._sign.value}{self._number}'

    @property
    def sign(self):
        return self._sign

    @property
    def number(self):
        return self._number


class Chromosome:
    def __init__(self, synteny_blocks):
        self._synteny_blocks = tuple(synteny_blocks)

    def __repr__(self):
        return f"({' '.join(map(str, self._synteny_blocks))})"

    def __iter__(self):
        yield from self._synteny_blocks


class Cycle:
    def __init__(self, nodes):
        self._nodes = tuple(nodes)

    def __repr__(self):
        return f"({' '.join(map(str, self._nodes))})"


def read_chromosome(line):
    line = line.replace('(', '').replace(')', '')
    synteny_blocks = []
    for item in line.split():
        signed_number = SignedNumber(Sign(item[0]), int(item[1:]))
        synteny_blocks.append(signed_number)
    return Chromosome(synteny_blocks)


def chromosome_to_cycle(chromosome):
    nodes = []
    for block in chromosome:
        if block.sign == Sign.PLUS:
            nodes.append(2*block.number-1)
            nodes.append(2*block.number)
        else:
            nodes.append(2*block.number)
            nodes.append(2*block.number-1)
    return Cycle(nodes)


def main():
    reader = sys.stdin
    chromosome = read_chromosome(next(reader))
    result = chromosome_to_cycle(chromosome)
    print(result)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter06/BA6F.scala
package TextbookTrack.Chapter06

object BA6F {
  sealed trait Sign
  final case object Plus extends Sign {
    override def toString: String = "+"
  }
  final case object Minus extends Sign {
    override def toString: String = "-"
  }

  object Sign {
    def apply(char: Char): Sign = char match {
      case '+' => Plus
      case '-' => Minus
      case _ => throw new Exception("Unknown sign.")
    }
  }

  final case class SignedNumber(sign: Sign, number: Int) {
    override def toString: String = s"$sign$number"
  }

  final case class Chromosome(syntenyBlocks: Vector[SignedNumber]) {
    override def toString: String = syntenyBlocks.mkString("(", " ", ")")

    def flatMap[T](f: SignedNumber => Vector[T]): Vector[T] = syntenyBlocks.flatMap(f)
  }

  final case class Cycle(nodes: Vector[Int]) {
    override def toString: String = nodes.mkString("(", " ", ")")
  }

  def readChromosome(line: String): Chromosome = {
    val syntenyBlocks: Vector[SignedNumber] =
      line
        .dropRight(1)
        .drop(1)
        .split(" ")
        .map{ item => SignedNumber(Sign(item(0)), item.drop(1).toInt) }
        .toVector
    Chromosome(syntenyBlocks)
  }

  def chromosomeToCycle(chromosome: Chromosome): Cycle = {
    val nodes: Vector[Int] = chromosome.flatMap{
      case SignedNumber(sign, number) =>
        if (sign == Plus) Vector(2*number - 1, 2*number)
        else Vector(2*number, 2*number - 1)
    }
    Cycle(nodes)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val chromosome: Chromosome = readChromosome(reader.next())
    val result: Cycle = chromosomeToCycle(chromosome)
    println(result)
  }
}
